### PR TITLE
Use potentialOSRPointHelper call as OSR point marker

### PR DESCRIPTION
**Pull request by liqunl**
This pull request contains three commits to use potentialOSRPointHelper
call as OSR point marker.

 - Insert a helper call right after each OSR point.
 - Remove the helper calls where OSR is not possible due to
    StringPeepHoles transformations.
 - Remove redundant helper calls in OSRGuardInsertion.

Details can be found in the commit message of each commit.

Closes: #3531

### CODE REVIEW

**Comment by liqunl** [comment]
Performance evaluation
1. Daytrader3

Relative throughput (compared to baseline), the higher the score, the better the performance

| Benchmark | Relative score| Standard deviation|
| ------------- | ------------- |------------- |
| daytrader3  | -0.1%  | 0.65%

With the change, we'll do HCRGuardAnalysis when there exists a bad yield point, which is the right thing to do but cost more compile time. The compile time overhead is 2.3% on daytrader3. The noise margin for compile time measurement is 0.43

Thus I have measured the performance on HiBench since it's compile time sensitive.

2. HiBench

HiBench measures the time spent to finish a task, so the smaller the score, the better the performance. 

| Benchmark | Relative score| Standard deviation|
| ------------- | ------------- |------------- |
| Pagerank  | -1.7%  |3.34%
| Terasort | -0.7% |2.28%
|Wordcount|+0.3%|1.56%|
|Sort|+0.3%|1.64%|

In conclusion, the change shows no regression on daytrader3 and HiBench

**Comment by liqunl** [comment]
@andrewcraik Can I have a review?

File: runtime/compiler/ilgen/Walker.cpp [hunk 1]
```diff
@@ -5478,17 +5499,15 @@ break
...
-       && comp()->isOSRTransitionTarget(TR::postExecutionOSR)
-       && comp()->isPotentialOSRPoint(callNodeTreeTop->getNode())
-       && !_methodSymbol->cannotAttemptOSRAt(callNode->getByteCodeInfo(), NULL, comp())
-       && !_cannotAttemptOSR)
+   if (needOSRBookkeeping)
       {
       saveStack(-1, !comp()->pendingPushLivenessDuringIlgen());
-      stashPendingPushLivenessForOSR(comp()->getOSRInductionOffset(callNode));
+      stashPendingPushLivenessForOSR(osrInductionOffset);
+      if (comp()->supportsInduceOSR())
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp]: I think you might want getOSRMode() == TR::voluntaryOSRMode to be sure we don't try and do this for involuntary OSR.

File: runtime/compiler/ilgen/Walker.cpp [hunk 2]
```diff
@@ -5323,6 +5323,27 @@ break
...
+   // The call may be transformed into a non-OSR point. Check if bookkeeping is needed
+   // before the transformation.
+   bool needOSRBookkeeping = false;
+   int32_t osrInductionOffset;
+
+   // callNodeTreeTop may be null if this call should not be placed in the trees
+   if (callNodeTreeTop
+       && comp()->getOption(TR_EnableOSR)
+       && !comp()->isPeekingMethod()
+       && comp()->isOSRTransitionTarget(TR::postExecutionOSR)
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp]: what about if the mode is preExecutionOSR do we need another helper marker for that?

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 3]
```diff
@@ -79,6 +79,53 @@
...
+      TR::TreeTop *tt = start;
+      do
+         {
+         TR::Node *osrNode = NULL;
+         if (comp()->isPotentialOSRPoint(tt->getNode(), &osrNode))
+            {
+            if (osrNode->isPotentialOSRPointHelperCall())
+               {
+               if (trace())
+                  traceMsg(comp(), "Remove tt n%dn with potential osr point %p n%dn\n", tt->getNode()->getGlobalIndex(), osrNode, osrNode->getGlobalIndex());
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: this should be a dumpOptDetails

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 4]
```diff
@@ -79,6 +79,53 @@
 #include "env/PersistentCHTable.hpp"           // for CLASSHASHTABLE_SIZE, etc
 #include "env/PersistentInfo.hpp"              // for PersistentInfo
 
+void TR_StringPeepholes::removePotentialOSRPointHelperCalls(TR::TreeTop* start, TR::TreeTop* end)
+   {
+   if (comp()->isOSRTransitionTarget(TR::postExecutionOSR) && comp()->getOption(TR_EnableOSR))
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: we should only do this for voluntary OSR modes

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 5]
```diff
@@ -79,6 +79,53 @@
...
+      // or on the surrounding trees.
+      //
+      TR::TreeTop *tt = start;
+      do
+         {
+         TR::Node *osrNode = NULL;
+         if (comp()->isPotentialOSRPoint(tt->getNode(), &osrNode))
+            {
+            if (trace())
+               traceMsg(comp(), "Can no longer OSR at [%p] n%dn\n", osrNode, osrNode->getGlobalIndex());
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: again dumpOptDetails rather than trace for this I think.

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 6]
```diff
@@ -309,7 +356,7 @@ void TR_StringPeepholes::processBlock(TR::Block *block)
    {
    TR::TreeTop *exit = block->getExit();
 
-   //traceMsg(comp(), "looking at block_%d\n", block->getNumber());
+   traceMsg(comp(), "looking at block_%d\n", block->getNumber());
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: please guard or clean up

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 7]
```diff
@@ -2092,9 +2143,25 @@ bool TR_StringPeepholes::checkMethodSignature(TR::SymbolReference *symRef, const
  */
 bool TR_StringPeepholes::skipNodeUnderOSR(TR::Node *node)
    {
-   return comp()->getOption(TR_EnableOSR)
-      && comp()->isOSRTransitionTarget(TR::postExecutionOSR)
-      && comp()->getMethodSymbol()->isOSRRelatedNode(node);
+   bool result = false;
+   if (comp()->getOption(TR_EnableOSR)
+      && comp()->isOSRTransitionTarget(TR::postExecutionOSR))
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: I don't know why the transition target affects this any OSR helper for a pre or post transitionw ould need to be skipable.

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 8]
```diff
@@ -79,6 +79,53 @@
...
+               TR::TransformUtil::removeTree(comp(), tt);
+               tt = prevTT;
+               }
+            }
+         tt = tt->getNextTreeTop();
+         } while (tt != end);
+      }
+   }
+
+void TR_StringPeepholes::fixUpOSRBookkeeping(TR::TreeTop* start, TR::TreeTop* end)
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: This is a bit misleading I think this needs to be something like prohibitOSROverRange or something like that since the bookkeeping is the stuff in the OMR layer tracking the live slots etc.

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 9]
```diff
@@ -79,6 +79,53 @@
...
+   {
+   if (comp()->isOSRTransitionTarget(TR::postExecutionOSR) && comp()->getOption(TR_EnableOSR))
+      {
+      // Walk region and ensure all possible OSR transition points has doNotProfile set on their bci
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: Can you please make this into doxygen for the method?

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 10]
```diff
@@ -332,7 +379,11 @@ void TR_StringPeepholes::processBlock(TR::Block *block)
 
             TR::TreeTop *newTree = detectPattern(block, tt, true);
             if (newTree)
+               {
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: I am concerned by the sequence which follows here. You are removing the OSR infrastructure before we have 'matched' the pattern and passed the point where we do the perform transformation to say that we are going to replace the peephole sequence with a new String of a private constructor targeting the specific sequence of appends we have found. This would be much more correct and much nicer if we skipped the OSR nodes as part of the pattern matching, ran the performTransformation and then did the remove of the OSR points before we did the manipulation to switch to the custom constructor. This pattern could see OSR infrastructure removed needlessly and could mean that doing a lastOptSubIndex in StringPeepholes could have unintended consequences on the OSR points and transitions seen by the OSR infrastructure which runs later.

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp [hunk 11]
```diff
@@ -30,63 +30,205 @@
...
+      if ((*guard)->getKind() == TR_HCRGuard || (*guard)->mergedWithHCRGuard())
+         return true;
+      }
 
-   // Currently, OSR guard insertion is only needed when in the OSR HCR mode
-   if (disableOSRGuards || !comp()->isOSRTransitionTarget(TR::postExecutionOSR))
+   return true;
+   }
+
+static bool hasYieldPointCannotOSR(TR::Compilation *comp)
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/OSRGuardInsertion.cpp]: I would prefer hasUnsupportedPotentialOSRPoint since this is about OSR points and the relationship to yield points is not determined - we are only concerned about transition points (which are normally yield points, but we don't need to conflate the two concepts).

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp [hunk 12]
```diff
@@ -30,63 +30,205 @@
...
+   bool needHCRGuardRemoval = hasHCRGuard(comp());
+   bool canInsertOSRGuards = !skipOSRGuardInsertion(comp());
+
+   if (canInsertOSRGuards && needHCRGuardRemoval)
       {
-      if (trace())
-         traceMsg(comp(), "No fear generating nodes - skipping\n");
-      comp()->getFlowGraph()->invalidateStructure();
-      return 0;
+      bool hasBadYieldPoint = hasYieldPointCannotOSR(comp());
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/OSRGuardInsertion.cpp]: hasBadYieldPoint should be something like hasUnsupportedOSRPoint or something like that...

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp [hunk 13]
```diff
@@ -30,63 +30,205 @@
...
+      // Future fear generating optimizations
+      //
+      if (!fearGeneratingNodes.isEmpty())
+         {
+         // Redudant potentialOSRPointHelper calls may result in more OSR guards than needed
+         //
+         if (hasBadYieldPoint)
+            removeRedundantPotentialOSRPointHelperCalls(guardAnalysis);
+         else
+            cleanUpPotentialOSRPointHelperCalls();
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/OSRGuardInsertion.cpp]: Why clean up here when there is a cleanup later on - this doesn't quite make sense to me... An explanatory comment is warranted.

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp [hunk 14]
```diff
@@ -30,63 +30,205 @@
...
+         treeTop = prevTreeTop;
+         }
       }
+   }
 
-   // Detect if there are no HCR guards
-   TR::list<TR_VirtualGuard*> &virtualGuards = comp()->getVirtualGuards();
-   auto guard = virtualGuards.begin();
-   for (; guard != virtualGuards.end(); ++guard)
+void TR_OSRGuardInsertion::removeRedundantPotentialOSRPointHelperCalls(TR_HCRGuardAnalysis* guardAnalysis)
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/OSRGuardInsertion.cpp]: Please provide a comment with an intuition for how this is working - I'm still working through the code to understand its interplay with the analyses in this opt pass.

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp [hunk 15]
```diff
@@ -30,63 +30,205 @@
...
+      TR::Node *osrNode = NULL;
+      if (comp()->isPotentialOSRPoint(ttNode, &osrNode))
+         {
+         if (visited.contains(osrNode))
+            continue;
+
+         if (protectedByOSRPoints &&
+             osrNode->isPotentialOSRPointHelperCall())
+            {
+            if (trace())
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/OSRGuardInsertion.cpp]: Is this optional? If so it should be a perform transformation. If not this is probably a dumpOptDetails kind of thing.

**Comment by andrewcraik** [comment]
I think some discussion of this change in relation to the pre and post OSR models is warranted - I think we need to be clear on when the helpers are needed and why since some fo the same issues affecting postExcecutionOSR could affect preExecutionOSR.

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp (hunk 11 above)
Comment (jdmpapin) [reply runtime/compiler/optimizer/OSRGuardInsertion.cpp]: While your suggested name would be more consistent with the naming of the queries in terms of which this is implemented, I think it's those queries that are confusing. "OSR point" sounds like a point at which OSR is possible, a transition point as you put it. An example of the confusion this causes: this function is not concerned with transition points at all. It is concerned instead with points that allow assumptions to be invalidated but which are specifically _not_ transition points. In the context of OSR specifically, I think the "yield point" terminology does make it clear which property is relevant, although it does conflict somewhat with the more general notion of a yield point, and with the existing queries `isPotentialOSRPoint*`.

Maybe we should try to move things consistently to a set of clearer terminology? e.g. abandon "OSR point" and speak only of "OSR transition point" and "OSR yield point"/"OSR invalidation point." Not as part of this PR, of course.

File: runtime/compiler/optimizer/StringPeepholes.cpp (hunk 10 above)
Comment (jdmpapin) [reply runtime/compiler/optimizer/StringPeepholes.cpp]: `detectPattern()` both matches the pattern and, if it finds a match, transforms the code - see the end of the function [starting here](https://github.com/eclipse/openj9/blob/7ebb4d4bef739fde5b4fcd6085f15449ffb25048/runtime/compiler/optimizer/StringPeepholes.cpp#L1863-L1890). I expect the other `detect*Pattern()` methods work similarly, considering that there doesn't seem to be any further transformation after skipping ahead to `newTree` in any of `processBlock`, `process`, or `perform` (except an attempt to inline some calls in the last of these).

Comment (liqunl) [reply runtime/compiler/optimizer/StringPeepholes.cpp]: Yes. all the `detect*Pattern()` methods detect and transform the code at the same time, then return the last tree in the pattern.

File: runtime/compiler/ilgen/Walker.cpp [hunk 16]
```diff
@@ -5380,6 +5380,27 @@ break
...
+       && comp()->isPotentialOSRPoint(callNodeTreeTop->getNode())
+       && !_methodSymbol->cannotAttemptOSRAt(callNode->getByteCodeInfo(), NULL, comp())
+       && !_cannotAttemptOSR)
+      {
+      needOSRBookkeeping = true;
+      // callNode may become a non-OSR point after the transformation, calling getOSRInductionOffset
+      // on a non-OSR point will trigger the assertion, thus get the induction offset here.
+      osrInductionOffset = comp()->getOSRInductionOffset(callNode);
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp]: Can you please share a case where the call node becomes something other than a call and we need the OSR point - ie we will still have a point where we want to decompile and that point is at an invalidation point? If it isn't an invalidation point why do we want the transition point?

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 17]
```diff
@@ -79,6 +79,80 @@
...
+ *     as they will not be able to reconstruct their stacks due to the optimizations done on them
+ *     or on the surrounding trees.
+ *
+ *  \param start
+ *     The tree marking the start of the range.
+ *
+ *  \param end
+ *     The tree marking the end of the range.
+ */
+void TR_StringPeepholes::prohibitOSROverRange(TR::TreeTop* start, TR::TreeTop* end)
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/StringPeepholes.cpp]: This also feels like a more general utility. I think it needs a check for voluntary OSR because it is not valid in involuntary OSR.

File: runtime/compiler/optimizer/StringPeepholes.cpp (hunk 7 above)
Comment (andrewcraik) [reply runtime/compiler/optimizer/StringPeepholes.cpp]: I think this needs a fast return for involuntary OSR where no skip is needed.

File: runtime/compiler/ilgen/Walker.cpp (hunk 16 above)
Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: `J9VMInternals.getSuperclass` will be transformed to a different call. A call to `JITHelpers.jitHelpers()` is also inserted. In this case, the invalidation point still exists but the calls are not OSR points. Even if the invalidation point is removed, there is no harm in doing the OSR bookkeeping and inserting the OSR helper call. Any redundant helper calls will be removed before inserting OSR guards.

**Comment by liqunl** [comment]
@andrewcraik could you review again?

File: runtime/compiler/ilgen/Walker.cpp (hunk 16 above)
Comment (andrewcraik) [reply runtime/compiler/ilgen/Walker.cpp]: right it just adds overhead, but given the perf numbers I guess we can live with that

**Comment by andrewcraik** [comment]
Ok I think this is looking good - I'm going to launch the builds and would just like a second opinion on correctness from @jdmpapin

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win.plinux jdk8,jdk11

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp [hunk 18]
```diff
@@ -30,63 +30,222 @@
...
+      //
+      TR_Structure *structure = requiresAnalysis ? fakeRegion(comp()) : NULL;
+      comp()->getFlowGraph()->setStructure(structure);
+
+      TR_HCRGuardAnalysis *guardAnalysis = requiresAnalysis ? new (comp()->allocator()) TR_HCRGuardAnalysis(comp(), optimizer(), structure) : NULL;
+      TR_BitVector fearGeneratingNodes(comp()->getNodeCount(), trMemory(), stackAlloc);
+
+      // Handle fear nodes for HCR guards, branch to slow path might be removed
+      //
+      if (needHCRGuardRemoval)
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/OSRGuardInsertion.cpp]: `needHCRGuardRemoval` is always true here. I think the logic is right as-is, but the `if` is misleading

**Comment by liqunl** [comment]
Failure is not related to this PR. There is a closed issue with the same test failure. https://github.com/eclipse/openj9/issues/594

File: runtime/compiler/optimizer/OSRGuardInsertion.cpp [hunk 19]
```diff
@@ -30,63 +30,222 @@
...
+   TR::list<TR_VirtualGuard*> &virtualGuards = comp->getVirtualGuards();
+   for (auto itr = virtualGuards.begin(), end = virtualGuards.end(); itr != end; ++itr)
+      {
+      if ((*itr)->getKind() == TR_HCRGuard || (*itr)->mergedWithHCRGuard())
+         return true;
+      }
 
-   // Currently, OSR guard insertion is only needed when in the OSR HCR mode
-   if (disableOSRGuards || !comp()->isOSRTransitionTarget(TR::postExecutionOSR))
+   return true;
```
[the remark this answers is not in the document]
Comment (liqunl) [reply runtime/compiler/optimizer/OSRGuardInsertion.cpp]: The query is `skipOSRGuardInsertion`

File: runtime/compiler/optimizer/StringPeepholes.cpp [hunk 20]
```diff
@@ -1945,9 +1964,9 @@ TR::TreeTop *TR_StringPeepholes::detectPattern(TR::Block *block, TR::TreeTop *tt
             // create a new treetop for String.valueOf()
             //                                                      nodeToConvert
             TR::SymbolReference * symRef = getSymRefTab()->findOrCreateMethodSymbol((appendedString[j] && appendedString[j]->getOpCode().hasSymbolReference()) ? appendedString[j]->getSymbolReference()->getOwningMethodIndex() : JITTED_METHOD_INDEX, -1, valueOfSymRef[j]->getSymbol()->getResolvedMethodSymbol()->getResolvedMethod(), TR::MethodSymbol::Static);
-            TR::Node *valueOf = TR::Node::createWithSymRef(TR::acall, 1, 1, appendedString[j], symRef);
+            TR::Node *valueOf = TR::Node::createWithSymRef(stringNode, TR::acall, 1, appendedString[j], symRef);
             TR::TreeTop::create(comp(), appendTree[j],
-                               TR::Node::create(TR::treetop, 1, valueOf));
+                               TR::Node::create(stringNode, TR::treetop, 1, valueOf));
```
[the remark this answers is not in the document]
Comment (liqunl) [reply runtime/compiler/optimizer/StringPeepholes.cpp]: The `append` calls are to be removed, the bci on `stringNode` will be used on the final tree, so I choose to use bci `stringNode`.

**Comment by jdmpapin** [comment]
> I believe we have some code somewhere that detects cases where the result of a call is immediately stored into a local, and moves the transition point for that call to after the store, so that the pending push will no longer be live at the transition. I couldn't easily find it, however.

@liqunl pointed me to [the code I was thinking of](https://github.com/eclipse/openj9/blob/fca3de3964302be44834593f7fe9c7705b0a1888/runtime/compiler/optimizer/OSRGuardInsertion.cpp#L378-L390), and it turns out I was mistaken. The target bytecode instruction for the transition is left unchanged, and the stores are repeated in the interpreter. In particular, the pending push temp holding the result of the call remains live. We simply move the OSR guard past repeatable stores so that splitting the block won't cause the call result to be stored and reloaded through yet *another* temp. This should just work for calls to `potentialOSRPointHelper`.

**Comment by liqunl** [comment]
WIP this PR as I need to add change to skip the helper call in StringBuilderTransformer

**Comment by liqunl** [comment]
@andrewcraik @jdmpapin Sorry for the last minute change. I added a change in StringBuilderTransformer to skip the helper call during pattern matching. Could you review again?

**Comment by liqunl** [comment]
@andrewcraik @jdmpapin ping?

**Comment by jdmpapin** [comment]
Still lgtm

**Comment by andrewcraik** [comment]
I think that looks reasonable - launching sanity again

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Comment by liqunl** [comment]
WIP this PR as it seems to have exposed a bug.
